### PR TITLE
More flexible rules for customizing docker images

### LINKED ISSUE #387: Supporting customization of tags in the built-in images
The operator does not currently support customizing the tags on the foundationdb and foundationdb-kubernetes-operator image. This can be a difficult limitation in some scenarios:

* Building one-off images for testing
* Building variant images, like a slim image and a full image

Allowing arbitrary tags will cause problems when doing upgrades, since the operator needs to carefully manage the image versions to upgrade the sidecar and the main container at the appropriate stages.

This came up in #342, and we addressed the problem by adding stronger validations. I think we should take another look at it and see if there's a cleaner way to support some customization of tags.

### PR DESCRIPTION
**Pull request by brownleej**
# Description

*Please include a summary of the change and which issue is addressed. If this change resolves an issue, please include the issue number in the description.*

This adds more options for customizing docker images, in order to cover more use cases like testing upgrades and having permanent image variants.

Resolves #387 

## Type of change

*Please select one of the options below.*

- New feature (non-breaking change which adds functionality)

# Discussion

*Are there any design details that you would like to discuss further?*

I considered deprecating the ability to specify custom images in the pod templates, since the new image configs can cover almost all the use cases for that, and I think it would lead to less confusion about how we form the images. However, that change wasn't covered in the original issue, and the pod templates allow customizing images on a per-process-class basis, so I did not want to deprecate support for that without further discussion.

This uses an implicit default tag suffix of `-1` for the sidecar image, and though it allows changing the suffix, it does not allow specifying a blank suffix. That matches the current behavior, but I don't like having defaults that people cannot replace with empty values. However, users can still specify a full tag as a workaround for this. I considered making the strings for these fields string pointers to allow us to distinguish between nil and an empty string, but that added significant complexity and didn't seem worth it to me.

# Testing

*Please describe the tests that you ran to verify your changes. Unit tests?
Manual testing?*

I added unit tests for how we form images with different variations on the image config, and created local clusters with different kinds of image names.

*Do we need to perform additional testing once this is merged, or perform in a larger testing environment?*

No.

# Documentation

*Did you update relevant documentation within this repository?*

Yes.

*If this change is adding new functionality, do we need to describe it in our user manual?*

Yes, I've added a section on this.

# Follow-up

*Are there any follow-up issues that we should pursue in the future?*

We may want an issue to discuss deprecating support for images in the pod spec, as I mentioned above. I also noticed in the course of this work that the deprecated fields in `ContainerOverrides` are not handled by the spec normalization, and are not caught in the deprecation checker, so we need to fix that before the 1.0 release.

*Does this introduce new defaults that we should re-evaluate in the future?*

No.

### CODE REVIEW

File: api/v1beta1/foundationdbcluster_types.go [hunk 1]
```diff
@@ -2084,6 +2071,52 @@ type ContainerOverrides struct {
 	SecurityContext *corev1.SecurityContext `json:"securityContext,omitempty"`
 }
 
+// ImageConfig provides a policy for customizing an image.
+type ImageConfig struct {
```
Comment (johscheuer) [inline api/v1beta1/foundationdbcluster_types.go:2075]: We should add some documentation here that the values are merged if multiple entries are matching.

File: internal/pod_models.go [hunk 2]
```diff
@@ -195,7 +184,7 @@ func GetPodSpec(cluster *fdbtypes.FoundationDBCluster, processClass fdbtypes.Pro
 		versionString = cluster.Spec.Version
 	}
 
-	image, err := GetImage(cluster.Spec.MainContainer.ImageName, mainContainer.Image, "foundationdb/foundationdb", versionString, processSettings.GetAllowTagOverride())
+	image, err := GetImage(mainContainer.Image, cluster.Spec.MainContainer.ImageConfigs, versionString, processSettings.GetAllowTagOverride())
```
Comment (johscheuer) [inline internal/pod_models.go:187]: Should we deprecate `processSettings.GetAllowTagOverride()` with this change?

**Review by johscheuer — APPROVED** [review]
One NIT otherwise LGTM 👍

File: internal/pod_models.go (hunk 2 above)
Comment (brownleej) [reply internal/pod_models.go:187]: I think we can talk about that in the context of deprecating custom images in the pod specs in general.

**Pull request merged**